apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfSubscriber.cs

```diff
@@ -462,55 +485,70 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
                             _byteCount += byteCount;
                             _elementCount += 1;
                             break;
                         }
+                        // FLow not OK, prepare to wait until a previous fn completes.
                         eventTask = _event.Task;
                     }
                     prevEventTask = eventTask;
                     await _taskHelper.ConfigureAwait(eventTask);
                 }
+                // Execute fn, and schedule the following code to execute once it has completed.
+                // Register the function, so we can be sure it's completed during shutdown.
                 _registerTaskFn(_taskHelper.Run(async () => await _taskHelper.ConfigureAwaitWithFinally(fn(), () =>
                 {
+                    TaskCompletionSource<int> ev = null;
                     lock (_lock)
                     {
                         bool preIsFlowOk = IsFlowOk();
                         _byteCount -= byteCount;
                         _elementCount -= 1;
                         if (!preIsFlowOk && IsFlowOk())
                         {
-                            var ev = _event;
+                            // If moving from flow-bad to flow-OK, then trigger the next execution.
+                            ev = _event;
                             _event = new TaskCompletionSource<int>();
-                            ev.SetResult(0);
                         }
                     }
+                    if (ev != null)
+                    {
+                        ev.SetResult(0);
+                    }
                 })));
             }
         }
 
+        /// <summary>
+        /// Controls a single <see cref="Channel"/>/<see cref="SubscriberClient"/> within this
+        /// <see cref="HiPerfSubscriberImpl"/>. This class controls the pulling of messages, and
+        /// the pushing of message acks and lease-extensions back to the server.
+        /// It also manages error conditions within the channel, restarting the StreamingPull()
+        /// RPC as required.
+        /// </summary>
         private class SingleChannel
         {
-            public SingleChannel(SubscriptionName subscriptionName,
+            public SingleChannel(HiPerfSubscriberImpl subscriber,
                 SubscriberClient client, Func<PubsubMessage, CancellationToken, Task<Reply>> handlerAsync,
-                TimeSpan autoExtendInterval, int modifyDeadlineSeconds, int maxAckExtendCount, Flow flow,
-                Action<Task> registerTaskFn, IScheduler scheduler, CancellationTokenSource globalSoftStopCts,
-                CancellationTokenSource globalHardStopCts, TaskHelper taskHelper)
+                Flow flow,
+                Action<Task> registerTaskFn)
             {
-                _subscriptionName = subscriptionName;
+                _subscriptionName = subscriber.SubscriptionName;
```

Comment:
I'd rather keep the fields. Otherwise all access has to indirect through a class/struct, which will make the code slightly messier.
Anyway, as you say, this can be changed later.